### PR TITLE
enhancement: improve apply to host flow

### PR DESCRIPTION
**Pull request by snitin315**
Resolve https://github.com/opencollective/opencollective/issues/4278

# Description

- Change the URL of the "Apply with a new Collective" flow and move it from https://opencollective.com/opensource/apply to https://opencollective.com/opensource/apply/new 

- Now that https://opencollective.com/opensource/apply is free, change its behavior and simply display the Collective Page with the "Apply" modal opened.

# Demo

https://user-images.githubusercontent.com/46647141/128299404-cd099bba-a6fa-476a-b296-31e5799b23bd.mp4

### CODE REVIEW

**Comment by snitin315** [comment]
> Now that https://opencollective.com/opensource/apply is free, change its behavior and simply display the Collective Page with the "Apply" modal opened.

@znarf  as [Open Collective Foundation](https://opencollective.com/foundation) has its own application flow, so this change doesn't apply to https://opencollective.com/foundation/apply

File: pages/collective-page.js [hunk 1]
```diff
@@ -178,6 +180,10 @@ class CollectivePage extends React.Component {
 
     const collective = data && data.Collective;
 
+    if (mode === 'apply' && !collective.isHost) {
+      return <Custom404 />;
+    }
+
```
Comment (znarf) [inline pages/collective-page.js:186]: Why is this needed? Code comment perhaps?

File: rewrites.js [hunk 2]
```diff
@@ -167,9 +167,14 @@ exports.REWRITES = [
   // New Create Collective Flow
   {
     source:
-      '/:hostCollectiveSlug?/:verb(apply|create)/:version(v2)?/:category(opensource|community|climate|covid-19)?/:step(form)?',
+      '/:hostCollectiveSlug?/:verb(apply/new|create)/:version(v2)?/:category(opensource|community|climate|covid-19)?/:step(form)?',
     destination: '/create-collective',
   },
+  // Apply to host
+  {
+    source: '/:slug/:mode(apply)',
```
Comment (znarf) [inline rewrites.js:175]: `action` instead of `mode` perhaps?

**Review by znarf — CHANGES_REQUESTED** [review]
It looks great!

Cherry on the cake, when someone click "Action -> Apply" in the menu, could you change the URL to `/apply`? That would be awesome.

File: pages/collective-page.js (hunk 1 above)
Comment (snitin315) [reply pages/collective-page.js:186]: As we don't want `/webpack/apply` to be a valid URL. I will add a comment.

File: rewrites.js (hunk 2 above)
Comment (snitin315) [reply rewrites.js:175]: `action` is better 👍🏻

**Comment by znarf** [comment]
> @znarf as Open Collective Foundation has its own application flow, so this change doesn't apply to https://opencollective.com/foundation/apply

While OCF has a custom flow, I think it should work similarly for OSC. On https://opencollective.com/foundation/apply we would open the modal too.

**Comment by snitin315** [comment]
> Cherry on the cake, when someone click "Action -> Apply" in the menu, could you change the URL to /apply

Done 👍🏻 

> While OCF has a custom flow, I think it should work similarly for OSC. On https://opencollective.com/foundation/apply we would open the modal too

Updated 👍🏻

File: rewrites.js [hunk 3]
```diff
@@ -15,7 +15,7 @@ exports.REWRITES = [
     destination: '/staticPage',
   },
   {
-    source: '/foundation/apply/:step(intro|fees|form|success)?',
+    source: '/foundation/apply/new/:step(intro|fees|form|success)?',
```
Comment (znarf) [inline rewrites.js:18]: Can we avoid introducing `new` here and simply send to a `step` (`intro` or `form`) when required?

File: rewrites.js [hunk 4]
```diff
@@ -167,9 +167,14 @@ exports.REWRITES = [
   // New Create Collective Flow
   {
     source:
-      '/:hostCollectiveSlug?/:verb(apply|create)/:version(v2)?/:category(opensource|community|climate|covid-19)?/:step(form)?',
+      '/:hostCollectiveSlug?/:verb(apply/new|create)/:version(v2)?/:category(opensource|community|climate|covid-19)?/:step(form)?',
```
Comment (znarf) [inline rewrites.js:170]: In the end, it's a bit weird to have `apply/new` as a parameter (having a slash in a parameter value doesn't look like a good practice). Can we use a different parameter. Put this in `:step` or introduce another parameter?

In the end maybe instead of `apply/new`, simply using `apply/form` is what we're looking for? I'm not sure, I might be missing something.

File: components/ocf-host-application/TermsOfFiscalSponsorship.js [hunk 5]
```diff
@@ -72,7 +72,7 @@ const TermsOfFiscalSponsorship = ({ checked, onChecked }) => (
         </Box>
       </Container>
     </Box>
-    <Link href="/foundation/apply/fees">
+    <Link href="/foundation/apply/new/fees">
```
Comment (znarf) [inline components/ocf-host-application/TermsOfFiscalSponsorship.js:75]: As state in other comments, if possible let's try avoiding introducing `new` in all routes. Having `fees` in the URL should be enough.

File: rewrites.js (hunk 3 above)
Comment (snitin315) [reply rewrites.js:18]: Yes we can avoid `new` here

File: rewrites.js (hunk 4 above)
Comment (snitin315) [reply rewrites.js:170]: using `apply/form` will directly lead to the `create-collective` form here, which should be the next step after clicking `request manual verification` here https://opencollective.com/opensource/apply/form 

Though moving `new` in `:step` will do the job.

File: components/ocf-host-application/TermsOfFiscalSponsorship.js (hunk 5 above)
Comment (snitin315) [reply components/ocf-host-application/TermsOfFiscalSponsorship.js:75]: Yes we can avoid it.

**Comment by snitin315** [comment]
Something strange, tests are passing locally but failing on CI, I'm looking into it.

File: rewrites.js (hunk 4 above)
Comment (snitin315) [reply rewrites.js]: moving `new` in `:step` wasn't working as expected, so I added `new` seperately-

`:verb(apply|create)/new/:version(v2)?`

Comment (snitin315) [reply rewrites.js]: Hmm, this can have side effects -

- `:verb(apply|create)` here will redirect `/apply` to `create-collective` instead of showing modal. 
-  Can't move `new` to `:step` or any other optional parameter as it will still redirect `/apply` to `/create-collective`
-  We can't introduce a new required parameter as `:verb(apply|create)/new/:version(v2)?` as we will have to update all URLs with `/create/new` `create/new/opensource` etc.


I think the safest option will be changing `/apply` to `/apply-new` to make `/apply` free - 

`:verb(apply-new|create)`

/cc @znarf

Comment (snitin315) [reply rewrites.js]: Or we can add a separate rewrite for `/apply/new` -

```js
{
    source: '/:hostCollectiveSlug?/apply/new',
    destination: '/create-collective',
  },
```

**Comment by znarf** [comment]
Thank you, will have a final look next Monday and tentatively merge.

**Review by znarf — CHANGES_REQUESTED** [review]
Testing right now, some feedback (will update as needed):

- [x] Polish: When closing the modal, the URL should change not not have `/apply` anymore
- [x] Polish: When opening the modal and using the browser back button, the modal is not closed. The route update should trigger a close.
- [x] Broken: The Open Source flow looks broken (when from `/apply-new` we click on "request manual verification") `/opensource/apply/form?hostTos=true` -> 404
- [x] Broken: The Foundation application process (with "Create new collective") is not going to the custom flow. Going to `/foundation/apply-new` instead of `https://opencollective.com/foundation/apply/intro`
- [x] Why have we introduced `/apply-new` instead of `/apply/new` ?

New:

- [ ] Polish: When opening the modal and using the browser back button, the modal is closed. Then, when using the forward browser button, the modal is not reopened.

File: components/ApplyToHostModal.js [hunk 6]
```diff
@@ -341,7 +341,7 @@ const ApplyToHostModal = ({ hostSlug, collective, onClose, onSuccess, router, ..
                               creatable
                               renderNewCollectiveOption={() => (
                                 <Link
-                                  href={isOCFHost ? `/foundation/apply/intro` : `/${host.slug}/apply`}
+                                  href={isOCFHost ? `/foundation/apply/intro` : `/${host.slug}/apply-new`}
```
Comment (snitin315) [inline components/ApplyToHostModal.js:344]: > The Foundation application process (with "Create new collective") is not going to the custom flow. Going to `/foundation/apply-new` instead of `https://opencollective.com/foundation/apply/intro`

This should be working as per logic here. `isOCFHost` is being calculated via `host.legacyId` which is specific for [`production` and `staging`](https://github.com/opencollective/opencollective-frontend/search?q=OPENCOLLECTIVE_FOUNDATION_ID) environment so this might not work in local env unless we use our local db id in the constant.

**Comment by snitin315** [comment]
> Why have we introduced /apply-new instead of /apply/new ?

Explained here https://github.com/opencollective/opencollective-frontend/pull/6724#discussion_r683973525

**Comment by snitin315** [comment]
Yes, I think we can simply use `/create`. I will test and update accordingly.

**Comment by snitin315** [comment]
@znarf I have updated the PR with the latest feedback 👍🏻 .

**Comment by znarf** [comment]
@snitin315 Thanks, everything works. Added (maybe) last one feedback in the list above.

**Review by znarf — CHANGES_REQUESTED** [review]
Looks good now but end to end test 29-host.dashboard.test.js is not passing. I restarted it and it definitely seems connected to the PR.

**Comment by snitin315** [comment]
Yes it's related I am looking into it.

**Comment by snitin315** [comment]
@znarf fixed 👍🏻 , the issue was that two modals were being inserted into DOM (one for hidden element)

**Review by znarf — APPROVED** [review]
Looks good.

One last thing, closing the apply modal doesn't play with routes perfectly well. Once you close it manually, you can't use the forward button to reopen it. If there is a simple solution, appreciated, otherwise it's already good!

**Comment by snitin315** [comment]
@znarf I did some research and tried some manipulation with `window.history` to achieve the behavior above but was unsuccessful, it was always having some side effects. It can introduce unnecessary complexity and may cause some unknown side effects.
 
I think we can create an issue for the same if anyone in the future may found or implement it with a simpler solution.
